### PR TITLE
Add additional DH groups from RFC 7919

### PR DESCRIPTION
**Pull request by andrewhop**
### Issues:
Addresses CryptoAlg-1874

### Description of changes: 
Add ffdhe4096 from RFC 7919 appendix 3. Add DH_new_by_nid which supports NID_ffdhe2048 and NID_ffdhe4096 from RFC 7919.

### Testing:
Added new tests that check the prime match as close as possible to the RFC, and that a KAT created with Sage math works.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and the ISC license.

### CODE REVIEW

**Review by dkostic — COMMENTED** [review]
it'd be good to add some tests for the new DH group

File: crypto/dh_extra/dh_test.cc [hunk 1]
```diff
@@ -300,33 +301,224 @@ TEST(DHTest, ASN1) {
...
+    EXPECT_TRUE(client_public);
+
+    const std::vector<uint8_t> server_secret_data = string_to_byte_array(test.server_sk);
+    bssl::UniquePtr<BIGNUM> server_secret(BN_bin2bn(server_secret_data.data(), server_secret_data.size(), nullptr));
+    EXPECT_TRUE(server_secret);
+
+    bssl::UniquePtr<DH> ffdhe2048_dh(DH_new_by_nid(test.nid));
+    EXPECT_TRUE(DH_set0_key(ffdhe2048_dh.get(), nullptr, server_secret.release()));
+    uint8_t buffer[4096];
+    int size = DH_compute_key(buffer,client_public.get(), ffdhe2048_dh.get());
```
Comment (dkostic) [inline crypto/dh_extra/dh_test.cc:517]:
```suggestion
    int size = DH_compute_key(buffer, client_public.get(), ffdhe2048_dh.get());
```
should we check `size` too?

Comment (andrewhop) [reply crypto/dh_extra/dh_test.cc:517]: That gets checked by Bytes's equality which relies on Span's equality that checks both are the same size. https://github.com/aws/aws-lc/blob/main/include/openssl/span.h#L42-L49

File: include/openssl/dh.h [hunk 2]
```diff
@@ -134,14 +138,18 @@ OPENSSL_EXPORT int DH_set0_pqg(DH *dh, BIGNUM *p, BIGNUM *q, BIGNUM *g);
 // |DH_generate_key| will use the bit length of p.
 OPENSSL_EXPORT int DH_set_length(DH *dh, unsigned priv_length);
 
-
```
Comment (samuel40791765) [inline include/openssl/dh.h:137]: I think the double new line might be intentional for `docs.go` to know this is a separate section

File: crypto/dh_extra/dh_test.cc [hunk 3]
```diff
@@ -300,33 +301,216 @@ TEST(DHTest, ASN1) {
...
+          "dc5c46f11c5ffe14152b9c7f9e504ae01ff84db158b9e48e9fbc46b99190cad2"
+          "e22113797dc7c81ad7c86bcd5e75405226459bb54b26fae179378e377ce5618e"
+          "65f04d2213e1991cefe991ec43272b6c7d93b51e2ccc3bf64486efd1e5c73b3b"
+          "f9344271ab9fbc43af41232ae7524c8213433ef39c64481e9cd06b9f9dc34226"
+          "85dfb2d69b8dc1af0f44d6f52d1d857ec28d93f459a23386ecfe5d97130e201d"
+          "90f159ff5995bbf766ceb38594b7a3192c99432e007b99f1ea7a828d15a1cba6"
+          "d86cd020ff64b1774cd35e33e3696a98574cedb64534f8ca88e2690709718d66"
+          "f4b88d759689819cde545d202b641b0529a02d588ff4c6b832c3f5a3d9bec9ec"
+          "ce0fb9af978b76bf93eba919c5bef844b4b1e2bff3d3758b577c70fa78d89a1d"
+           "d5a1864a2d3795c3668562c67aa77265f38812f001d28b25f7965109481ec2c7")
```
Comment (samuel40791765) [inline crypto/dh_extra/dh_test.cc:496]: Very very minor nit :)
```suggestion
          "d5a1864a2d3795c3668562c67aa77265f38812f001d28b25f7965109481ec2c7")
```

File: crypto/dh_extra/dh_test.cc [hunk 4]
```diff
@@ -300,33 +301,216 @@ TEST(DHTest, ASN1) {
...
 
-  uint8_t buffer[sizeof(kPrime1536)];
-  ASSERT_EQ(BN_num_bytes(bn.get()), sizeof(kPrime1536));
-  ASSERT_EQ(BN_bn2bin(bn.get(), buffer), sizeof(kPrime1536));
-  EXPECT_EQ(Bytes(buffer), Bytes(kPrime1536));
+    bssl::UniquePtr<DH> ffdhe2048_dh(DH_new_by_nid(test.nid));
+    EXPECT_TRUE(DH_set0_key(ffdhe2048_dh.get(), nullptr, server_secret.release()));
+    uint8_t buffer[4096];
+    int size = DH_compute_key(buffer, client_public.get(), ffdhe2048_dh.get());
+    EXPECT_LE(size, 4096);
```
Comment (samuel40791765) [inline crypto/dh_extra/dh_test.cc:510]: `DH_compute_key` returns a negative number on error, so any errors might fall through the gaps here. Although the next test would still fail so..

**Review by dkostic — APPROVED** [review]

**Review by samuel40791765 — APPROVED** [review]

**Pull request merged**